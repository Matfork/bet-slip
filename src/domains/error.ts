export interface ErrorResponse {
  status: number;
  message: string;
}

export class ApiError extends Error {
  status: number;

  constructor({ message, status }: { message: string; status: number }) {
    super(message); // (1)
    this.status = status; // (2)
  }
}
