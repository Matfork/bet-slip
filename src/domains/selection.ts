export enum SelectedSelectionOptions {
  Select,
  Unselect,
}

export interface SelectionType {
  id: string;
  name: string;
  price: number;
}
