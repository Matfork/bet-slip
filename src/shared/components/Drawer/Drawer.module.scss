@import "src/assets/scss/spacing";

$card-width: 180px;
$card-height: $card-width - 40px;

.Drawer-component__container {
  padding: $spacing-3;

  .Drawer-component__card {
    padding: $spacing-2;
    display: flex;
    justify-content: center;
    align-items: center;
    width: $card-width;
    height: $card-height;
    margin-bottom: $spacing-3;
    text-align: center;
    flex-direction: column;
  }
}
