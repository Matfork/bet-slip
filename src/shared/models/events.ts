export interface EventType {
  id: string;
  name: string;
  markets: MarketType[];
}

interface MarketType {
  id: string;
  name: string;
  markets: SelectionType[];
}

interface SelectionType {
  id: string;
  name: string;
  price: number;
}
