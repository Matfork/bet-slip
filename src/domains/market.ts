import { SelectionType } from './selection';

export interface MarketType {
  id: string;
  name: string;
  selections: SelectionType[];
}
