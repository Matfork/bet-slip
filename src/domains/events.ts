import { MarketType } from './market';

export interface EventType {
  id: string;
  name: string;
  markets: MarketType[];
}
