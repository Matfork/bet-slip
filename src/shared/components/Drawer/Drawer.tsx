import React from 'react';
import { Button, Card, Drawer, Typography } from '@material-ui/core';
import { RootState } from 'store/reducers';
import { EventsInitialState } from 'store/reducers/events.reducer';
import { useDispatch, useSelector } from 'react-redux';
import { SelectionType, SelectedSelectionOptions } from 'domains/selection';
import { EventsActionCreators } from 'store/actions/action-creators/events.actionCreator';

import './Drawer.scss';

interface DrawerProps {
  showDrawer: boolean;
  onToggleDrawer: (arg: boolean) => void;
}

const className = 'Drawer-component';

const DrawerComponent: React.FC<DrawerProps> = ({ showDrawer, onToggleDrawer }: DrawerProps) => {
  const dispatch = useDispatch();
  const { userSelection } = useSelector<RootState, EventsInitialState>((state) => state.events);

  const handleRemoveSelection = (selection: SelectionType): void => {
    dispatch({
      type: EventsActionCreators.SetUserSelection,
      payload: {
        selection,
        option: SelectedSelectionOptions.Unselect,
      },
    });
  };

  return (
    <Drawer anchor="right" open={showDrawer} onClose={() => onToggleDrawer(false)} className={className}>
      <div className={`${className}__container`}>
        {userSelection.length ? (
          userSelection.map((selection) => (
            <Card variant="outlined" key={selection.id} className={`${className}__card`}>
              <Typography variant="subtitle1">{selection.name}</Typography>
              <Typography variant="subtitle2">{selection.price}</Typography>
              <Button color="primary" variant="contained" onClick={() => handleRemoveSelection(selection)}>
                Delete
              </Button>
            </Card>
          ))
        ) : (
          <Card variant="outlined" className={`${className}__card`}>
            <Typography variant="body1">You don&apos;t have any bet selected, try selecting one option first! </Typography>
          </Card>
        )}
      </div>
    </Drawer>
  );
};

export default DrawerComponent;
