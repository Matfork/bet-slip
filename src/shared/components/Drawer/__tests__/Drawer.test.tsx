import { fireEvent, screen } from '@testing-library/react';
import { customRender } from 'test.utilities';
import { initialStateStub, selectionsStub } from 'test.stub';
import { EventsActionCreators } from 'store/actions/action-creators/events.actionCreator';
import { SelectedSelectionOptions } from 'domains/selection';
import Drawer from '../Drawer';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...(jest.requireActual('react-redux') as {}),
  useDispatch: () => mockDispatch,
}));

describe('Drawer Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should test snapshot', () => {
    const component = customRender(<Drawer showDrawer onToggleDrawer={jest.fn()} />);
    expect(component.container).toMatchSnapshot();
  });

  it('should not display drawer if showDrawer is set to false', () => {
    const component = customRender(<Drawer showDrawer={false} onToggleDrawer={jest.fn()} />);

    const drawerContainer = component.container.querySelector('.Drawer-component__container');
    expect(drawerContainer).not.toBeInTheDocument();
  });

  it('should display drawer if showDrawer is set to true', () => {
    customRender(<Drawer showDrawer onToggleDrawer={jest.fn()} />);

    const titleEl = screen.queryByText(/have any bet selected, try selecting one option/);
    expect(titleEl).toBeInTheDocument();
  });

  it('should display no data available', () => {
    customRender(<Drawer showDrawer onToggleDrawer={jest.fn()} />);

    const text = screen.queryByText(/have any bet selected, try selecting one option/)!;
    expect(text).toBeInTheDocument();
  });

  it('should display data available based on user selection', () => {
    customRender(<Drawer showDrawer onToggleDrawer={jest.fn()} />, {
      initialState: {
        ...initialStateStub,
        events: { ...initialStateStub.events, userSelection: selectionsStub.slice(0, 2) },
      },
    });

    let text = screen.queryByText(/Real Madrid/)!;
    expect(text).toBeInTheDocument();

    text = screen.queryByText(/Barcelona/)!;
    expect(text).toBeInTheDocument();
  });

  it('should delete userSelection on button clicked', () => {
    customRender(<Drawer showDrawer onToggleDrawer={jest.fn()} />, {
      initialState: {
        ...initialStateStub,
        events: { ...initialStateStub.events, userSelection: [selectionsStub[0]] },
      },
    });

    const button = screen.queryByText(/Delete/)?.closest('button')!;
    fireEvent.click(button);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: EventsActionCreators.SetUserSelection,
      payload: {
        selection: selectionsStub[0],
        option: SelectedSelectionOptions.Unselect,
      },
    });
  });

  it('should call onToggleDrawer on drawer closed action', async () => {
    const onToggleDrawer = jest.fn();
    const component = customRender(<Drawer showDrawer onToggleDrawer={onToggleDrawer} />);

    const titleEl = screen.queryByText(/have any bet selected, try selecting one option/);
    expect(titleEl).toBeInTheDocument();

    const divBackdrop = component.container.closest('body')!.querySelector('.MuiBackdrop-root')!;
    fireEvent.click(divBackdrop);

    expect(onToggleDrawer).toHaveBeenCalled();
  });
});
