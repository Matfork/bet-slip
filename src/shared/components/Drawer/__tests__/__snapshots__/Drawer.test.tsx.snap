// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Component should test snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
